### ISSUE TITLE
Proposal: Add traces_filter

### ISSUE THREAD

**Issue by st0012** [getsentry/sentry-ruby#1470]
This callback is for filtering out a **sampled transaction** with information that's not accessible during sampling, for example:

- The response's format and status code.
- Scope data added during the request (tags, user...etc.)?
  - In Rails applications, the controller's transaction name (`PostsController#index`).

## Use Cases 

- Skip transactions for based on status (e.g. 404 or 30x requests). https://github.com/getsentry/sentry-ruby/issues/1252
- Skip transactions for not wanted response format.
  - In Rails applications, it's not always possible to predict the response format early in the request.
  - Currently, sentry-rails uses path prefix matching to workaround this: https://github.com/getsentry/sentry-ruby/blob/master/sentry-rails/lib/sentry/rails/capture_exceptions.rb#L37-L41
- Skip transactions with information that's only available after sampling, like controller's transaction name. https://github.com/getsentry/sentry-ruby/issues/1429

### It's NOT for

- Adding/removing data from the transaction.

### Triggered Timing

It should be called inside `Transaction#finish`, right before the transactions is sent.

This has several benefits:

- Users don't need to learn where to place this.
- At that moment, the transaction should already have every information users need.

### Input

- The transaction object.
- Some information from the current scope, like user, tags and transaction name.
  - In Rails applications, transaction name will change after the request enters a controller.
  - See https://github.com/getsentry/sentry-ruby/issues/1429
- Parent sampled?

#### Transaction

- It must be a sampled transaction.
- It should already contains the status code.

### Output

- It only returns `true` or `false`/`nil` to decide whether to keep the transaction.

### Other Behavior

- It logs result similar to what `Transaction#set_initial_sampling_decision` does.
- When failed, it logs message similar to `Client#send_event` and raise the exception.

### Example

```ruby

# the current traces_sampler example, as a comparison
config.traces_sampler = lambda do |sampling_context|
  unless sampling_context[:parent_sampled].nil?
    next sampling_context[:parent_sampled]
  end

  transaction_context = sampling_context[:transaction_context]

  op = transaction_context[:op]
  transaction_name = transaction_context[:name]

  case op
  when /request/
    case transaction_name
    when /health_check/
      0.0
    when /payment/
      0.5
    when /api/
      0.2
    else
      0.1
    end
  when /sidekiq/
    0.01 # you may want to set a lower rate for background jobs if the number is large
  else
    0.0 # ignore all other transactions
  end
end

config.traces_filter do |filter_context|
  transaction = filter_context[:transaction_cotnext]
  scope_context = filter_context[:scope_context] # may contain info like user, tags or transaction
  
  op = transaction_context[:op]

  case op
  when /request/
    case scope_context[:transaction_name]
    when /HelloController/ # skips all actions of HelloController
      next false # in Ruby we use next to early return block
    end
    
    case transaction_context[:status]
    when "not_found", "permission_denied", "unauthenticated"
      next false
    end
  end
  
  true
end
```

## Other Notes

- Since this is an experimental feature to all SDKs, perhaps we can add a tag to transactions that went through this callbak? This will help us analysis how users adopts this feature.

**Comment by st0012** [comment]
@rhcarvalho @HazAT I'd like to hear what your think about this 🙂

**Comment by rhcarvalho** [comment]
It sounds too similar to what `EventProcessor` can do today to solve those use cases.

We're currently working on simplifying the tracing API, among other things de-emphasizing `Transaction` and, thus, wouldn't add more API around that concept.

**Comment by st0012** [comment]
If we can get rid of most of the sampling, that'd be the best 👍 
Is there an issue/discussion..etc. that I can learn more about this plan?

**Comment by rhcarvalho** [comment]
Not yet.

**Comment by st0012** [comment]
I see. closing this now then.

**Comment by vitobotta** [comment]
@st0012 Awesome, It would be great if you could update this issue then :) Thanks!

**Comment by st0012** [comment]
@vitobotta the fix has been released in version 4.6.5 🙂
